# MATH 210 Introduction to Mathematical Computing

## Project 1

In [1]:
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
import scipy.optimize as spi
import numpy.linalg as la

In [2]:
spi.linprog?

## The following are the documentation of `scipy.optimize.linprog`

Signature: spi.linprog(c, A_ub=None, b_ub=None, A_eq=None, b_eq=None, bounds=None, method='simplex', callback=None, options=None)
Docstring:
Minimize a linear objective function subject to linear
equality and inequality constraints.

Linear Programming is intended to solve the following problem form:

Minimize:     c^T * x

Subject to:   A_ub * x <= b_ub
              A_eq * x == b_eq

Parameters
----------
A `scipy.optimize.OptimizeInput` consisting of the following fields:
    
    c : array_like
        Coefficients of the linear objective function to be minimized.
    A_ub : array_like, optional
        2-D array which, when matrix-multiplied by x, gives the values of the
        upper-bound inequality constraints at x.
    b_ub : array_like, optional
        1-D array of values representing the upper-bound of each inequality
        constraint (row) in A_ub.
    A_eq : array_like, optional
        2-D array which, when matrix-multiplied by x, gives the values of the
        equality constraints at x.
    b_eq : array_like, optional
        1-D array of values representing the RHS of each equality constraint
        (row) in A_eq.
    bounds : sequence, optional
        ``(min, max)`` pairs for each element in ``x``, defining
        the bounds on that parameter. Use None for one of ``min`` or
        ``max`` when there is no bound in that direction. By default
        bounds are ``(0, None)`` (non-negative)
        If a sequence containing a single tuple is provided, then ``min`` and
        ``max`` will be applied to all variables in the problem.
    method : str, optional
        Type of solver.  At this time only 'simplex' is supported
        :ref:`(see here) <optimize.linprog-simplex>`.
    callback : callable, optional
        If a callback function is provide, it will be called within each
        iteration of the simplex algorithm. The callback must have the signature
        `callback(xk, **kwargs)` where xk is the current solution vector
        and kwargs is a dictionary containing the following::

            "tableau" : The current Simplex algorithm tableau
            "nit" : The current iteration.
            "pivot" : The pivot (row, column) used for the next iteration.
            "phase" : Whether the algorithm is in Phase 1 or Phase 2.
            "basis" : The indices of the columns of the basic variables.

    options : dict, optional
        A dictionary of solver options. All methods accept the following
        generic options:

            maxiter : int
                Maximum number of iterations to perform.
            disp : bool
                Set to True to print convergence messages.

        For method-specific options, see `show_options('linprog')`.

Returns
-------
A `scipy.optimize.OptimizeResult` consisting of the following fields:

    x : ndarray
        The independent variable vector which optimizes the linear
        programming problem.
    fun : float
        Value of the objective function.
    slack : ndarray
        The values of the slack variables.  Each slack variable corresponds
        to an inequality constraint.  If the slack is zero, then the
        corresponding constraint is active.
    success : bool
        Returns True if the algorithm succeeded in finding an optimal
        solution.
    status : int
        An integer representing the exit status of the optimization::

             0 : Optimization terminated successfully
             1 : Iteration limit reached
             2 : Problem appears to be infeasible
             3 : Problem appears to be unbounded

    nit : int
        The number of iterations performed.
    message : str
        A string descriptor of the exit status of the optimization.

See Also
--------
show_options : Additional options accepted by the solvers

Notes
-----
This section describes the available solvers that can be selected by the
'method' parameter. The default method is :ref:`Simplex <optimize.linprog-simplex>`.

Method *Simplex* uses the Simplex algorithm (as it relates to Linear
Programming, NOT the Nelder-Mead Simplex) [1]_, [2]_. This algorithm
should be reasonably reliable and fast.

.. versionadded:: 0.15.0

References
----------
.. [1] Dantzig, George B., Linear programming and extensions. Rand
       Corporation Research Study Princeton Univ. Press, Princeton, NJ, 1963
.. [2] Hillier, S.H. and Lieberman, G.J. (1995), "Introduction to
       Mathematical Programming", McGraw-Hill, Chapter 4.
.. [3] Bland, Robert G. New finite pivoting rules for the simplex method.
       Mathematics of Operations Research (2), 1977: pp. 103-107.

Examples
--------
Consider the following problem:

Minimize: f = -1*x[0] + 4*x[1]

Subject to: -3*x[0] + 1*x[1] <= 6
             1*x[0] + 2*x[1] <= 4
                        x[1] >= -3

where:  -inf <= x[0] <= inf

This problem deviates from the standard linear programming problem.
In standard form, linear programming problems assume the variables x are
non-negative.  Since the variables don't have standard bounds where
0 <= x <= inf, the bounds of the variables must be explicitly set.

There are two upper-bound constraints, which can be expressed as

dot(A_ub, x) <= b_ub

The input for this problem is as follows:

>>> c = [-1, 4]
>>> A = [[-3, 1], [1, 2]]
>>> b = [6, 4]
>>> x0_bounds = (None, None)
>>> x1_bounds = (-3, None)
>>> from scipy.optimize import linprog
>>> res = linprog(c, A_ub=A, b_ub=b, bounds=(x0_bounds, x1_bounds),
...               options={"disp": True})
Optimization terminated successfully.
     Current function value: -22.000000
     Iterations: 1
>>> print(res)
     fun: -22.0
 message: 'Optimization terminated successfully.'
     nit: 1
   slack: array([ 39.,   0.])
  status: 0
 success: True
       x: array([ 10.,  -3.])

Note the actual objective value is 11.428571.  In this case we minimized
the negative of the objective function.
File:      /opt/conda/lib/python3.5/site-packages/scipy/optimize/_linprog.py
Type:      function

## Introduction
The package I gonna use is `scipy.optimize.linprog`, which is a subpackage of `scipy.optimize`. It is designed to solve the linear programing problem using the so called `Simplex` method. I learned this method also in MATH 340 last semester and I thought it is a interesting topic and sufficient way to solve it with computer (we can solve it by hand but just with small amount of variables). We used software called `Lindo` to deal with the problem when it came to a large amount of variables and statistic analysis. So this time, I will try use Python to solve the problem.

`Linear Programming` might best be called `Linear Optimization`: it means finding maxima and minima of linear functions of several variables subject to constraints that are linear equations or linear inequalities. The word “programming” has the old-fashioned meaning of “planning” and was chosen in the forties, before the advent of computers.

The `Linear Programming problem` can be use to solve the question such as entering an industry or not for a company, how to allocate resources to maximize the profits, etc.

## Learning outcomes

The main learning outcome is to let people who read this notebook know how to use `scipy.optimize.linprog` to deal with the linear programming problem.

Furthermore, I will give some simple examples to show how it works in the real world problem. Actually I tried to find some firms' annual report to see whether they properly allocate their resources to maximize the profit, but it seems most resources are classified and I cannot give the whole linear system.

## Standard form
The standard form of this kind of problem is:

$$
max \ \ c_1x_1 + c_2x_2 + \cdots + c_nx_n
$$
$$
s.t \ \ a_{11}x_1 + a_{12}x_2 + \cdots + a_{1n}x_n \leq b_1 
\\ a_{21}x_1 + a_{22}x_2 + \cdots + a_{2n}x_n \leq b_2
\\ \cdots \\
\\ a_{m1}x_1 + a_{m2}x_2 + \cdots + a_{mn}x_n \leq b_m
$$
$$
x_1, \cdots ,x_n \geq 0
$$

## Matrix formula
Also, we usually write the standard form in matrix formula:

$$
max \ \ c^Tx
$$
$$
s.t \ \ 
\begin{pmatrix} a_{11} & \cdots & a_{1n} \\
a_{21} & \cdots & a_{2n} \\
\vdots & \vdots & \vdots \\
a_{m1} & \cdots & a_{mn} \end{pmatrix}
\begin{pmatrix} x_1 \\ \vdots \\ x_n \end{pmatrix} \ \ \leq \ \
\begin{pmatrix} b_1 \\ \vdots \\ b_m \end{pmatrix}
\\ x \geq 0
$$
where 
$$
c = \begin{pmatrix} c_1 \\ \vdots \\ c_n \end{pmatrix}
$$

The function you are trying to maximize or minimize is called the `objective function`. Each of the inequalities or equations the variables must satisfy is called a `constraint`. Constraint that simply specify that a variable is non-negative, such as $c \geq 0$, are called `positivity constraints`. We’ll almost always assume that each variable has a positivity constraint and the Simplex Method relies on this.

## Simplex Method

The way to solve this kind of problem is called [`Simplex Method`](http://www.math.ubc.ca/~oantolin/math340/simplex-intro.html). The procedure is:
* Transform all equations into standard form
* If any b < 0, introduce new variable $x_0$ and do the `Phase 1` which make sure everything is in standard form
* Now everything is feasible and do the `Phase 2` to solve the linear system


## Simple Example
 
Let me introduce a simple example with `scipy.optimize.linprog`:

$$
maximize \ \ x_1 + 2x_2 - x_3 \\
s.t. \ \ 2x_1 + x_2 + x_3 \leq 14 \\
         4x_1 + 2x_2 + 3x_3 \leq 28 \\
         2x_1 + 5x_2 + 5x_3 \leq 30 \\
         x_1 \ , x_2 \ , x_3 \geq 0
$$

In [5]:
A = np.array([[2,1,1],[4,2,3],[2,5,5]])
spi.linprog([-1,-2,1],A,[14,28,30])

     fun: -13.0
 message: 'Optimization terminated successfully.'
     nit: 2
   slack: array([ 0.,  0.,  0.])
  status: 0
 success: True
       x: array([ 5.,  4.,  0.])

Since the `scipy.optimize.linprog` always solve the `minimize` problem, so we convert our object function into minimize $-cx$. So the result above shows that the optimal solution to our linear programming is 13 with $x_1 = 5 \ , x_2 = 4 \ , x_3 = 0$ and slack variables $x_4 = x_5 = x_6 = 0$.

## Tollbooth Staffing Problem (Limitation of the method)

The tollbooth staffing problem is a interesting application of linear programming.

The problem is to minimize the total number of the tollbooth staff under the following constraints:
  * every tollbooth requires one staff
  * every shift is a 9-hour shift including a 1-hour rest (ie. 4-hour work - 1-hour rest - 4-hour work)
  * $x_i$ is the number of works on the i-th shift starting at midnight + $(i-1)$hrs
    * e.g. $x_6$ <--> 5am - 9am, 10am - 2pm
  * number of workers' constraints:
    * between midnight and 6am, 2 people should be available
    * between 6am and 10am, 8 people should be available
    * between 10am and 12pm, 4 people should be available
    * between 12pm and 4pm, 3 people should be available
    * between 4pm and 6pm, 6 people should be available
    * between 10pm and midnight, 3 people should be available

Then we can transform all these information into a linear system:
$$
min \ \
\begin{align}
& x_1+x_2+x_3+x_4+x_5+x_6+x_7+x_8+x_9+\\
& x_{10}+x_{11}+x_{12}+x_{13}+x_{14}+x_{15}+x_{16}+x_{17}+\\
& x_{18}+x_{19}+x_{20}+x_{21}+x_{22}+x_{23}+x_{24}
\end{align}
$$
$$
\begin{align}
  & subject \ \ to \\
  & 12:00am) \ \ x_1+x_{17}+x_{18}+x_{19}+x_{20}+x_{22}+x_{23}+x_{24} > 2 \\
  & 1:00am) \ \ x_1+ x_2+x_{18}+x_{19}+x_{20}+x_{21}+x_{23}+x_{24} > 2 \\
  & 2:00am) \ \ x_1+ x_2+ x_3+x_{19}+x_{20}+x_{21}+x_{22}+x_{24} > 2 \\
  & 3:00am) \ \ x_1+ x_2+ x_3+ x_4+x_{20}+x_{21}+x_{22}+x_{23} > 2 \\
  & 4:00am) \ \ x_2+ x_3+ x_4+ x_5+x_{21}+x_{22}+x_{23}+x_{24} > 2 \\
  & 5:00am) \ \ x_1+ x_3+ x_4+ x_5+ x_6+x_{22}+x_{23}+x_{24} > 2 \\
  & 6:00am) \ \ x_1+ x_2+ x_4+ x_5+ x_6+ x_7+x_{23}+x_{24} > 8 \\
  & 7:00am) \ \ x_1+ x_2+ x_3+ x_5+ x_6+ x_7+ x_8+x_{24} > 8 \\
  & 8:00am) \ \ x_1+ x_2+ x_3+ x_4+ x_6+ x_7+ x_8+ x_{9} > 8 \\
  & 9:00am) \ \ x_2+ x_3+ x_4+ x_5+ x_7+ x_8+ x_{9}+x_{10} > 8 \\
  & 10:00am) \ \ x_3+ x_4+ x_5+ x_6+ x_8+ x_{9}+x_{10}+x_{11} > 4 \\
  & 11:00am) \ \ x_4+ x_5+ x_6+ x_7+ x_{9}+x_{10}+x_{11}+x_{12} > 4 \\
  & 12:00pm) \ \ x_5+ x_6+ x_7+ x_8+x_{10}+x_{11}+x_{12}+x_{13} > 3 \\
  & 1:00pm) \ \ x_6+ x_7+ x_8+ x_{9}+x_{11}+x_{12}+x_{13}+x_{14} > 3 \\
  & 2:00pm) \ \ x_7+ x_8+ x_{9}+x_{10}+x_{12}+x_{13}+x_{14}+x_{15} > 3 \\
  & 3:00pm) \ \ x_8+ x_{9}+x_{10}+x_{11}+x_{13}+x_{14}+x_{15}+x_{16} > 3 \\
  & 4:00pm) \ \ x_{9}+x_{10}+x_{11}+x_{12}+x_{14}+x_{15}+x_{16}+x_{17} > 6 \\
  & 5:00pm) \ \ x_{10}+x_{11}+x_{12}+x_{13}+x_{15}+x_{16}+x_{17}+x_{18} > 6 \\
  & 6:00pm) \ \ x_{11}+x_{12}+x_{13}+x_{14}+x_{16}+x_{17}+x_{18}+x_{19} > 5 \\
  & 7:00pm) \ \ x_{12}+x_{13}+x_{14}+x_{15}+x_{17}+x_{18}+x_{19}+x_{20} > 5 \\
  & 8:00pm) \ \ x_{13}+x_{14}+x_{15}+x_{16}+x_{18}+x_{19}+x_{20}+x_{21} > 5 \\
  & 9:00pm) \ \ x_{14}+x_{15}+x_{16}+x_{17}+x_{19}+x_{20}+x_{21}+x_{22} > 5 \\
  & 10:00pm) \ \ x_{15}+x_{16}+x_{17}+x_{18}+x_{20}+x_{21}+x_{22}+x_{23} > 3 \\
  & 11:00pm) \ \ x_{16}+x_{17}+x_{18}+x_{19}+x_{21}+x_{22}+x_{23}+x_{24} > 3 \\
\end{align}
$$

And then we can transform it into `Python Language` and solve it with `scipy.optimize.linprog`.

In [3]:
A = np.array([[-1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,-1,-1,-1,-1,-1,-1,-1],
              [-1,-1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,-1,-1,-1,-1,-1,-1],
              [-1,-1,-1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,-1,-1,-1,-1,-1],
              [-1,-1,-1,-1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,-1,-1,-1,-1],
              [0,-1,-1,-1,-1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,-1,-1,-1,-1],
              [-1,0,-1,-1,-1,-1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,-1,-1,-1],
              [-1,-1,0,-1,-1,-1,-1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,-1,-1],
              [-1,-1,-1,0,-1,-1,-1,-1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,-1],
              [-1,-1,-1,-1,0,-1,-1,-1,-1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],
              [0,-1,-1,-1,-1,0,-1,-1,-1,-1,0,0,0,0,0,0,0,0,0,0,0,0,0,0],
              [0,0,-1,-1,-1,-1,0,-1,-1,-1,-1,0,0,0,0,0,0,0,0,0,0,0,0,0],
              [0,0,0,-1,-1,-1,-1,0,-1,-1,-1,-1,0,0,0,0,0,0,0,0,0,0,0,0],
              [0,0,0,0,-1,-1,-1,-1,0,-1,-1,-1,-1,0,0,0,0,0,0,0,0,0,0,0],
              [0,0,0,0,0,-1,-1,-1,-1,0,-1,-1,-1,-1,0,0,0,0,0,0,0,0,0,0],
              [0,0,0,0,0,0,-1,-1,-1,-1,0,-1,-1,-1,-1,0,0,0,0,0,0,0,0,0],
              [0,0,0,0,0,0,0,-1,-1,-1,-1,0,-1,-1,-1,-1,0,0,0,0,0,0,0,0],
              [0,0,0,0,0,0,0,0,-1,-1,-1,-1,0,-1,-1,-1,-1,0,0,0,0,0,0,0],
              [0,0,0,0,0,0,0,0,0,-1,-1,-1,-1,0,-1,-1,-1,-1,0,0,0,0,0,0],
              [0,0,0,0,0,0,0,0,0,0,-1,-1,-1,-1,0,-1,-1,-1,-1,0,0,0,0,0],
              [0,0,0,0,0,0,0,0,0,0,0,-1,-1,-1,-1,0,-1,-1,-1,-1,0,0,0,0],
              [0,0,0,0,0,0,0,0,0,0,0,0,-1,-1,-1,-1,0,-1,-1,-1,-1,0,0,0],
              [0,0,0,0,0,0,0,0,0,0,0,0,0,-1,-1,-1,-1,0,-1,-1,-1,-1,0,0],
              [0,0,0,0,0,0,0,0,0,0,0,0,0,0,-1,-1,-1,-1,0,-1,-1,-1,-1,0],
              [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,-1,-1,-1,-1,0,-1,-1,-1,-1]])
c = np.array([1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1])
b = np.array([-2,-2,-2,-2,-2,-2,-8,-8,-8,-8,-4,-4,-3,-3,-3,-3,-6,-6,-5,-5,-5,-5,-3,-3])
spi.linprog(c,A,b)

     fun: 15.874999999999993
 message: 'Optimization terminated successfully.'
     nit: 41
   slack: array([ 0.   ,  2.625,  3.   ,  3.375,  4.25 ,  1.5  ,  0.   ,  0.   ,
        0.   ,  0.   ,  0.   ,  0.375,  0.   ,  0.125,  0.75 ,  1.5  ,
        0.   ,  0.   ,  0.   ,  0.   ,  0.   ,  0.   ,  3.   ,  1.5  ])
  status: 0
 success: True
       x: array([ 0.5  ,  4.125,  0.375,  0.375,  1.375,  0.875,  0.75 ,  0.   ,
        1.   ,  0.   ,  0.   ,  0.   ,  0.   ,  0.5  ,  1.5  ,  1.5  ,
        1.5  ,  1.5  ,  0.   ,  0.   ,  0.   ,  0.   ,  0.   ,  0.   ])

So from the above result we can get the solution with

$$
x_8 = x_{10} = x_{11} = x_{12} = x_{13} = x_{19} = x_{20} = x_{21} = x_{22} = x_{23} = x_{24} = 0 \\ 
x_2 = 4.125 \\ 
x_3 = x_4 = 0.375 \\ 
x_5 = 1.375 \\ 
x_6 = 0.875 \\
x_1 = x_{14} = 0.5 \\
x_7 =  0.75 \\
x_9 = 1 \\
x_{15} = x_{16} = x_{17} = x_{18} = 1.5 
$$

But as we all know, we can not have say 4.125 people at $x_2$. So it will involve the integer programming problem. But the default method for `scipy.optimize.linprog` is `Simplex` and for now it can only support `Simplex` method. Thus, this kind of question can not be solved with `scipy.optimize.linprog` and it is kind of the limitation.

## The Career Points Problem

A student wants to take on extra curricular activities that will boost their chance of finding a job after graduation. We quantify the boost in terms of career points. There are limits on the students time, money and loss of self esteem. The activities of networking, taking professional exams and doing sports will all contribute to landing a job after graduation; each unit of the activities will result in some number of career points. The student wishes to maximize the number of career points. The activities (for this student) have different costs in terms of time, money and self esteem. The student has 34 hours to spend on these activities and has `$1800` to spend (the costs are given in units of `$100` below). A student can only lose 9 units of self esteem before affecting their success in UBC courses.


|               | networking | exams | sports | available |
|---------------|------------|-------|--------|-----------|
| hours         | 5          | 6     | 4      | 34        |
| money         | 3          | 3     | 2      | 18        |
| self esteem   | 1          | 2     | 1      | 9         |
| career points | 7          | 8     | 5      |           |

Let $x_1$ denote the units of networking, $x_2$ denote the units of professional exams and $x_3$ denote the units of sports. And therefore we can obtain the following data:

$$
A = \begin{pmatrix} 5 & 6 & 4 \\ 3 & 3 & 2 \\ 1 & 2 & 1 \end{pmatrix} \ ,
x = \begin{pmatrix} x_1 \\ x_2 \\ x_3 \end{pmatrix} \ ,
b = \begin{pmatrix} 34 \\ 18 \\ 9 \end{pmatrix} \ ,
c = \begin{pmatrix} 7 & 8 & 5 \end{pmatrix}
$$

In [4]:
A = np.array([[5,6,4],[3,3,2],[1,2,1]])
b = np.array([34,18,9])
c = np.array([-7,-8,-5])
spi.linprog(c,A,b)

     fun: -45.0
 message: 'Optimization terminated successfully.'
     nit: 2
   slack: array([ 1.,  0.,  0.])
  status: 0
 success: True
       x: array([ 3.,  3.,  0.])

As the result above, we obtain the solution with 3 units of networking and 3 units of exams but no sports. But as the matter of fact that we can also get the solution with 2 units of networking, 2 units of exams and 1 unit of sports, which gives us the same optimum value. That means the optimum solution to certain problems are sometimes not single.

## The Construction Problem

We wish to create water supply for a small community at minimum total cost. We can construct roof collections sytems, wells and dams. We are not concerned with integrality in this question. We seek a total supply of water of 100 litres per hour. We have a limited availability of maintenance time which is required for wells and dams. As well, we impose a maximum on the number of each water system.

|              | roof | well | dam |    constraint    |
|:------------:|:----:|:----:|:---:|:----------------:|
|    supply    |   1  |   5  |  50 | total $\geq$ 100 |
| maintainance |   0  |   1  |  10 |  total $\leq$ 21 |
|     upper    |  20  |  10  |  2  |                  |

Then we can obtain:

$$
A = \begin{pmatrix} -1 & -5 & -50 \\ 0 & 1 & 10 \\ 1 & 0 & 0 \\ 0 & 1 & 0 \\ 0 & 0 & 1 \end{pmatrix} \ ,
x = \begin{pmatrix} x_1 \\ x_2 \\ x_3 \end{pmatrix} \ ,
b = \begin{pmatrix} -100 \\ 21 \\ 20 \\ 10 \\ 2 \end{pmatrix} \ ,
c = \begin{pmatrix} -10 \\ -30 \\ -300 \end{pmatrix}
$$

In [6]:
A = np.array([[-1,-5,-50],[0,1,10],[1,0,0],[0,1,0],[0,0,1]])
b = np.array([-100,21,20,10,2])
c = np.array([10,30,300])
spi.linprog(c,A,b)

     fun: 600.0
 message: 'Optimization terminated successfully.'
     nit: 1
   slack: array([  1.,  20.,  10.,   0.,   0.])
  status: 0
 success: True
       x: array([ 0.,  0.,  2.])

From the above result, we get the optimum solution with 2 units of dam and nothing else. But also, this can also be obtain by 10 units of well and 1 unit of dam.

## Conclusion

So it seems not that difficult to deal with the linear programming problem with `scipy.optimize.linprog`, we just need to pay attention to the following:

* First transfering all the things into the standard form
* The standard form is to maximize the object function but minimize in the `scipy.optimize.linprog`, so we need to times `-1` to the standard form
* Then constructing right matrix A, b and c
* Use `scipy.optimize.linprg(c,A,b)` to compute

## Reference

Omar Antolín Camarena, my instructor for [MATH 340](http://www.math.ubc.ca/~oantolin/math340/) last term

Wikipedia page of [Linear Programming](https://en.wikipedia.org/wiki/Linear_programming)

`Microeconomics`, Pindyck, Robert and Rubinfeld, David, Prentice Hall, 8-th edition